### DISCUSSION TITLE
Support for compound keys?

### DISCUSSION THREAD

**Discussion by chrisbartley** [directus/directus#7721] [Archive]
[First, a _huge_ thank you for Directus!  It's really quite amazing--fast, powerful, intuitive, and super easy to use! 👏 ]

Say I have two (MySQL) tables, Categories and Subcategories, with the following fields:

Categories
- id
- name

Subcategories
- id
- name
- category_id

Category names are required to be unique, but not so for Subcategories (e.g.  there might be two "Movement" subcategories, where one is related to the "Input" category, and the other related to the "Output" category).  But I'd really like to enforce, at the database level, that the _pairing_ of subcategory name and category_id is unique.  This is easy enough in MySQL with compound keys, and I can create the Subcategories table like this:

```sql
create table subcategories
(
   id          int unsigned auto_increment primary key,
   name        varchar(255) not null,
   category_id int unsigned not null,
   constraint subcategories_name_category_id_unique_key
      unique (name, category_id),
   constraint subcategories_categories_id_fk
      foreign key (category_id) references categories (id)
);

create unique index subcategories_id_uindex
	on subcategories (id);

create index subcategories_name_index
	on subcategories (name);
```

I was amazed to discover that I can execute the above SQL, then pop over to the Directus admin and it sees the new `subcategories` table and prompts me to click it to configure.  Neat!

But, doing so, is where things go kinda haywire. The interface is duplicating the `category_id` and `id` fields:

![image](https://user-images.githubusercontent.com/86031/131528670-69d26602-4024-4aae-adbe-b4e347b0cf59.png)

And then attempting to reorder fields, by dragging that first `id` field to the top, I get _lots_ of duplicates (it's in the interface only...nothing has actually changed at the MySQL level):

![image](https://user-images.githubusercontent.com/86031/131528806-aaf14952-6f28-49fc-b06c-54ccd7b0a8c2.png)

Have I just set things up wrong?  Should Directus be able to work with compound keys?

**Comment by rijkvanzanten** [comment]
> Should Directus be able to work with compound keys?

No, Directus isn't made to handle compound primary keys at this point..

Which brings me to an interesting question: How would you imagine it would work ideally? It's definitely an interesting feature request, and something we could use in the system tables as well (eg compound key on collection+field for directus_fields), but I've always had some trouble trying to imagine how it'd work in things like API routing (f.e. `/items/subcategories/5+1` ? )

**Reply by chrisbartley** [reply]
Ah, good point...I hadn't gone that far with my thinking :-)  I'd be fine with compound, non-primary keys, if that makes things simpler.  Either way I would end up implementing the constraint in business logic code, so was mostly just exploring whether I could also enforce it at the database level.  Thanks, and thanks for the super-fast reply!

**Reply by rijkvanzanten** [reply]
Ohh that's another good point. Supporting compound keys doesn't necessarily have to mean supporting them as the _primary_ key of an item. That does in fact simplify things a little bit 🤔 

I believe the duplicate fields is a known bug around fields that have more than one index/constraint associated with it, so that could solve the wonkiness in the api/app itself. I'm also curious what sort of UX would be best to configure a compound key 🤔 Would you have a separate section under fields for compound keys, or maybe have a "Make the current field unique _with_ these other fields?" sort of thing on the individual fields? Interesting questions!
